File: pyrrm-gui/frontend/src/types/analysis.ts
import type { PlotlyFigure } from './index'

export interface SessionSummary {
  id: string
  name: string
  folder_path: string
  gauge_ids: string[]
  total_experiments: number
  total_failures: number
  loaded_at: string
}

export interface GaugeMetadata {
  area_km2: number | null
  gauge_name: string | null
  record_start: string | null
  record_end: string | null
  record_years: number | null
  n_days: number | null
  mean_precip_mm_day: number | null
  mean_pet_mm_day: number | null
  total_precip_mm_yr: number | null
  total_pet_mm_yr: number | null
  mean_flow: number | null
  median_flow: number | null
  aridity_index: number | null
  runoff_ratio: number | null
}

export interface GaugeSummary {
  gauge_id: string
  n_experiments: number
  n_failures: number
  best_by_objective: Record<string, { key: string; value: number | null }>
  metadata: GaugeMetadata | null
}

export interface GeoLayer {
  name: string
  available: boolean
  n_features: number
}

export interface GeoLayersResponse {
  layers: GeoLayer[]
}

export interface SessionDetail extends SessionSummary {
  gauges: GaugeSummary[]
}

export interface ExperimentInfo {
  key: string
  model: string
  objective: string
  algorithm: string
  transformation: string | null
  best_objective: number
  runtime_seconds: number | null
  success: boolean
  headline_metrics: Record<string, number | null>
}

export interface DiagnosticsRow {
  experiment_key: string
  metrics: Record<string, number | null>
}

export interface ClustermapData {
  heatmap_values: (number | null)[][]
  row_labels: string[]
  col_labels: string[]
  row_dendrogram: DendrogramData
  col_dendrogram: DendrogramData
  annotations: (string | null)[][]
}

export interface DendrogramData {
  icoord: number[][]
  dcoord: number[][]
  leaves: number[]
}

export interface DiagnosticsResponse {
  raw_table: DiagnosticsRow[]
  normalised_table: DiagnosticsRow[]
  clustermap: ClustermapData | null
  top_experiments: { experiment_key: string; mean_score: number }[]
  metric_groups: Record<string, string[]>
}

export interface SummaryTableRow {
  key: string
  gauge_id: string
  model: string
  objective: string
  algorithm: string
  transformation: string | null
  best_objective: number | null
  runtime_seconds: number | null
  success: boolean
  parameters: Record<string, number | null>
}

// ── Signature comparison types ──────────────────────────────────────────────

export interface SignatureCategoryData {
  signatures: string[]
  observed: Record<string, number | null>
  experiments: Record<string, Record<string, number | null>>
  percent_errors: Record<string, Record<string, number | null>>
}

export interface SignatureComparisonResponse {
  categories: Record<string, SignatureCategoryData>
  bar_figures: Record<string, PlotlyFigure>
  heatmap_figure: PlotlyFigure | null
}

// ── Signature reference types ───────────────────────────────────────────────

export interface SignatureInfo {
  id: string
  name: string
  category: string
  units: string
  range: (number | string)[] | null
  description: string
  formula: string
  interpretation: string
  related: string[]
  references: string[]
}

export interface SignatureReferenceResponse {
  categories: Record<string, SignatureInfo[]>
  category_order: string[]
  total_signatures: number
}

// ── Report card types ────────────────────────────────────────────────────────

export interface ReportCardHeader {
  catchment_name: string
  gauge_id: string
  area: string | number | null
  method: string
  objective_name: string
  best_objective: number
  period_start: string
  period_end: string
}

export interface ReportCardResponse {
  header: ReportCardHeader
  hydrograph_linear: PlotlyFigure
  hydrograph_log: PlotlyFigure
  metrics_table: PlotlyFigure
  fdc: PlotlyFigure
  scatter: PlotlyFigure
  parameters: PlotlyFigure | null
  signatures_table: PlotlyFigure | null
}

export type { PlotlyFigure }

File: pyrrm-gui/backend/app/schemas/analysis.py
"""
Pydantic schemas for the Batch Analysis API.
"""

from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field


class LoadSessionRequest(BaseModel):
    """Request to load batch results from a folder."""
    folder_path: str = Field(
        ...,
        description="Absolute path to folder containing batch_result.pkl files"
    )
    name: Optional[str] = Field(
        None,
        description="Human-readable session name; defaults to folder basename"
    )


class SessionSummary(BaseModel):
    """Summary of a loaded analysis session."""
    id: str
    name: str
    folder_path: str
    gauge_ids: List[str]
    total_experiments: int
    total_failures: int
    loaded_at: str


class GaugeMetadata(BaseModel):
    """Catchment / gauge physical metadata extracted from CalibrationReports."""
    area_km2: Optional[float] = None
    gauge_name: Optional[str] = None
    record_start: Optional[str] = None
    record_end: Optional[str] = None
    record_years: Optional[float] = None
    n_days: Optional[int] = None
    mean_precip_mm_day: Optional[float] = None
    mean_pet_mm_day: Optional[float] = None
    total_precip_mm_yr: Optional[float] = None
    total_pet_mm_yr: Optional[float] = None
    mean_flow: Optional[float] = None
    median_flow: Optional[float] = None
    aridity_index: Optional[float] = None
    runoff_ratio: Optional[float] = None


class GaugeSummary(BaseModel):
    """Summary of a single gauge within a session."""
    gauge_id: str
    n_experiments: int
    n_failures: int
    best_by_objective: Dict[str, Dict[str, Any]]
    metadata: Optional[GaugeMetadata] = None


class ExperimentInfo(BaseModel):
    """Parsed experiment information with headline metrics."""
    key: str
    model: str
    objective: str
    algorithm: str
    transformation: Optional[str] = None
    best_objective: float
    runtime_seconds: Optional[float] = None
    success: bool = True
    headline_metrics: Dict[str, Optional[float]] = Field(default_factory=dict)


class DiagnosticsRow(BaseModel):
    """Single row of the diagnostics table (one experiment)."""
    experiment_key: str
    metrics: Dict[str, Optional[float]]


class ClustermapData(BaseModel):
    """Data needed to render a clustermap in Plotly."""
    heatmap_values: List[List[Optional[float]]]
    row_labels: List[str]
    col_labels: List[str]
    row_dendrogram: Dict[str, Any]
    col_dendrogram: Dict[str, Any]
    annotations: List[List[Optional[str]]]


class DiagnosticsResponse(BaseModel):
    """Full diagnostics response for a gauge."""
    raw_table: List[DiagnosticsRow]
    normalised_table: List[DiagnosticsRow]
    clustermap: ClustermapData
    top_experiments: List[Dict[str, Any]]
    metric_groups: Dict[str, List[str]]


class SessionDetailResponse(BaseModel):
    """Detailed session info."""
    id: str
    name: str
    folder_path: str
    gauge_ids: List[str]
    total_experiments: int
    total_failures: int
    loaded_at: str
    gauges: List[GaugeSummary]


class SummaryTableRow(BaseModel):
    """Row in the combined summary table."""
    key: str
    gauge_id: str
    model: str
    objective: str
    algorithm: str
    transformation: Optional[str] = None
    best_objective: Optional[float] = None
    runtime_seconds: Optional[float] = None
    success: bool = True
    parameters: Dict[str, Optional[float]] = Field(default_factory=dict)
